### PR TITLE
Improve MethodHandle direct dispatch J2I-prevention transformations

### PR DESCRIPTION
**Pull request by jdmpapin**
...by avoiding breaking the block containing the original call. There isn't much opportunity to optimize the inline logic anyway, so it's better to keep the IL neater.

On code generators that support doing so (initially just x86-64), calls to `linkToStatic`, `linkToSpecial`, and `invokeBasic` that have not been refined are now changed in-place into calls to a new JIT non-helper `<jitDispatchJ9Method>`, which takes the callee `J9Method` as an additional (first) argument. It has JIT private linkage, but the extra argument is treated specially so that the other arguments are passed in exactly the way expected by (any JIT-compiled body for) the actual callee. The code generator generates a sequence that determines whether the callee is compiled and jumps to the JIT entry point if it is. Otherwise, it jumps out of line to do a J2I transition.

If the environment variable `TR_stressJitDispatchJ9MethodJ2I` is set, then the back end generates an unconditional jump to the J2I path, exercising it even when the callee is compiled.

----

There is one additional commit to get rid of the `mov edi, <imm>` instruction in the call snippet when calling an OSR induction helper on x86-32, mostly for consistency.

----

This depends on eclipse/omr#7090

### CODE REVIEW

File: runtime/codert_vm/thunkcrt.c [hunk 1]
```diff
@@ -55,10 +55,10 @@
 /* Encoding format is:
  *
  * 1 byte			argCount
- * 128 bytes		encoded bytes (up to 255 args + 1 return type)
+ * 129 bytes		encoded bytes (up to 1 receiver + 255 args + 1 return type)
  */
 
-#define J9_THUNK_MAX_ENCODED_BYTES	129
+#define J9_THUNK_MAX_ENCODED_BYTES	130
```
Comment (jdmpapin) [inline runtime/codert_vm/thunkcrt.c:61]: Looking at this again, I'm not sure that I need to adjust this value. The receiver should be included in the 255-parameter limit in the JVM

Comment (jdmpapin) [reply runtime/codert_vm/thunkcrt.c:61]: Yes, the receiver contributes toward the argument count for the purpose of the limit of 255. JVMS section 4.3.3:
> A method descriptor is valid only if it represents method parameters with a total length of 255 or less, where that length includes the contribution for `this` in the case of instance or interface method invocations.

So I'll change this back

**Comment by jdmpapin** [comment]
Rebased to fix a conflict

**Comment by 0xdaryl** [comment]
@gacholio : could you review the proposed changes in codert_vm please?

**Comment by gacholio** [comment]
Why is the new assembly helper only on X, when there are clearly changes on other architectures here?

**Comment by gacholio** [comment]
> * Make J2I thunks usable for static methods

I'm not sure this is valid - the stack walker distinguishes virtual and static calls. I'll have to look at the walker and transition code to be sure if this is bad or not.

Is this an attempt to save the generated code for backstoring arguments in static dispatch snippets?

**Comment by jdmpapin** [comment]
> Is this an attempt to save the generated code for backstoring arguments in static dispatch snippets?

I wrote it this way because originally I had moved all of the logic out to assembly, and I had in mind the way we do the direct dispatches from PicBuilder. In that setting it didn't occur to me to jump back from the helper into the JIT body to run a snippet. But then it turned out that it was slow to have the JIT-to-JIT path out of line, so I inlined it

I was writing a long reply answering about other architectures and detailing why I think the strategy in this PR should work
<details>
<summary>(click for very abridged version if interested)</summary>
<ul>
<li>The changes on other platforms were only to adapt to the change in the API of the J2I thunk management functions
<li>I think this strategy should work because the virtual and static J2I paths in the VM seem to differ only in how they determine the callee <tt>J9Method</tt>, and because when the JIT generates a direct call to an interpreted method, we already generate the call in the same way for both virtual and static callees, so the stack walker can't expect the compiled code to look different in those two cases
</ul>
</details>

but as I was writing that, I saw how this change could be much less invasive if I use (almost) the same interpreter call snippet we use for direct calls. So I'll try that. Thanks @gacholio for the nudge

**Comment by jdmpapin** [comment]
Updated to leave J2I thunks alone and instead use the call snippet

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk11,jdk21 depends eclipse/omr#7090

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk11,jdk21 depends eclipse/omr#7090

Previous CI was a resounding failure.  Restarting.

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional,sanity.openjdk all jdk8,jdk11,jdk17,jdk21 depends eclipse/omr#7090

**Comment by pshipton** [comment]
I killed that PR build, it's likely to take out jenkins. Pls avoid testing the unb platforms: amac, xlinux, xmac. I've disabled all the unb plinux so that platform should be ok, although we have limited machines at osu.

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional win,alinux,plinux,zlinux jdk17 depends eclipse/omr#7090

**Comment by 0xdaryl** [comment]
I'm satisfied that a sufficient coverage of tests have passed on this and that the effort to manually shepherd the jobs failing for infrastructure reasons is not a productive use of time. That, coupled with Devin's own internal testing on this, leads me to conclude that this is safe to merge.  I will do a coordinated merge early in the morning barring a reasonable objection @pshipton @hzongaro @jdmpapin .

**Comment by jdmpapin** [comment]
I don't believe a coordinated merge is necessary for this change. It should be fine to merge eclipse/omr#7090 first (but please check that you agree)

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**